src/org/jitsi/impl/neomedia/transform/fec/FlexFec03Mask.java

```diff
@@ -0,0 +1,239 @@
+    private static final int MASK_0_END_BIT = 15;
+    private static final int MASK_1_K_BIT = 16;
+    private static final int MASK_1_START_BIT = 17;
+    private static final int MASK_1_END_BIT = 47;
+    private static final int MASK_2_K_BIT = 48;
+    private static final int MASK_2_START_BIT = 49;
+    private static final int MASK_2_END_BIT = 111;
+
+    private static final int MASK_SIZE_SMALL = 2;
+    private static final int MASK_SIZE_MED = 6;
+    private static final int MASK_SIZE_LARGE = 14;
+
+    private int sizeBytes;
+    /**
+     * The mask field (including k bits)
+     */
+    private LeftToRightBitSet maskWithKBits;
+    private int baseSeqNum;
+
+    /**
+     * Initialize this maskWithoutKBits with a received buffer.
+     * buffer + maskOffset should point to the location of the start
+     * of the mask
+     * @param buffer the flexfec packet buffer
+     * @param maskOffset maskOffset to the location of the start of the mask
+     * @param baseSeqNum the base sequence number from the flexfec packet
+     */
+    FlexFec03Mask(byte[] buffer, int maskOffset, int baseSeqNum)
+    {
+        this.sizeBytes = getMaskSizeInBytes(buffer, maskOffset);
+        this.maskWithKBits = LeftToRightBitSet.valueOf(buffer, maskOffset, this.sizeBytes);
+        this.baseSeqNum = baseSeqNum;
+    }
+
+    /**
+     * Create a mask from a base sequence number and a list of protected
+     * sequence numbers
+     * @param baseSeqNum the base sequence number to use for the mask
+     * @param protectedSeqNums the sequence numbers this mask should mark
+     * as protected
+     */
+    public FlexFec03Mask(int baseSeqNum, List<Integer> protectedSeqNums)
+    {
+        this.sizeBytes = getMaskSizeInBytes(baseSeqNum, protectedSeqNums);
+        this.baseSeqNum = baseSeqNum;
+        this.maskWithKBits = createMaskWithKBits(this.sizeBytes, this.baseSeqNum, protectedSeqNums);
+    }
+
+    private static LeftToRightBitSet createMaskWithKBits(int sizeBytes, int baseSeqNum, List<Integer> protectedSeqNums)
+    {
+        int numBits = sizeBytes * 8;
+        LeftToRightBitSet mask = new LeftToRightBitSet(numBits);
+        // First create a mask without the k bits
+        for (Integer protectedSeqNum : protectedSeqNums)
+        {
+            int delta = protectedSeqNum - baseSeqNum;
+            mask.set(delta);
+        }
+
+        // Shift the mask bits
+        if (sizeBytes > MASK_SIZE_SMALL)
+        {
+            if (sizeBytes > MASK_SIZE_MED)
+            {
+                // Shift to make room for the k bit
+                // We subtract one from the mask start constants here since they
+                // include the k bit
+                LeftToRightBitSet.shiftBitsRight(mask, MASK_2_START_BIT - 3, numBits - 1, 1);
+            }
+            LeftToRightBitSet.shiftBitsRight(mask, MASK_1_START_BIT - 2, numBits - 1, 1);
+        }
+        LeftToRightBitSet.shiftBitsRight(mask, MASK_0_START_BIT - 1, numBits - 1, 1);
```

Comment:
I really hope the format changes for the final spec. :)

Reply:
a lot of nastiness tucked away in this file :)